libs/execution_context/ebpf_program.c

```diff
@@ -21,6 +21,23 @@ static size_t _ebpf_program_state_index = MAXUINT64;
 #define EBPF_MAX_HASH_SIZE 128
 #define EBPF_HASH_ALGORITHM L"SHA256"
 
+// The maximum number of iterations to run when executing a BPF_PROG_RUN before yielding to the scheduler. This is to
+// prevent starving other threads on the CPU, but also allows the measurement of the time spent in the BPF program
+// to be reasonably accurate.
+// Assumptions:
+// 1) 2Ghz processor, with worst case of 2 cycles per instruction (taking into account cache misses).
+// 2) Approximately 4 CPU instructions per BPF instruction.
+// 3) Maximum program size of 4K instructions.
```

Comment:
Agreed, it's not number of instructions, its maximum executed instruction count, which is what the verifier enforces.

My only counter argument is this:
How many packets does the NIC indicate up as part of a single DPC? Assuming the BPF program runs for each packet, then the same argument should apply. 

As a side note:
As per @dylandreimerink
> From what I understand (not an scheduling expert), we disable normal premption but switch into a co-operative mode. Linux checks if the scheduler has requested that we reschedule, and if that is the case, we pause the timer, reschedule and resume the timer once we get control back.

I don't think Windows has anything like this though.

Reply:
MSFT sources suggest that chains of packets up to 128 aren't uncommon, so 64 invocations isnt unreasonable. [NDIS APIs](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ndis/nf-ndis-ndismindicatereceivenetbufferlists) (i.e. XDP path) don't place a limit on the number of packets that can be indicated.

Reply:
Actually, it occurred to me that it could be worse:
bpf_tail_call can permit chaining up to 32 programs together, so if verifier enforces 100000 and there can be up to 32 programs, that means it could be as high as 3.2 million.

Reply:
That's true.  Hence my initial intuition that 1 invocation per DPC is best.

Reply:
Switching to KeShouldYieldProcessor and eliminating the heuristic.